### ISSUE TITLE
Why is Doorkeeper returning expired authorization codes?

### ISSUE THREAD

**Issue by theycallmeswift** [doorkeeper-gem/doorkeeper#882]
I've got an OAuth API using Doorkeeper 3.1.0.  For a small number of users (~5%) when they request an authorization code, the API is returning an expired or revoked code.

For example, here's a line from today's server logs (2016-09-17):

```
2016-09-17T19:36:30.032406+00:00 heroku[router]: at=info method=GET path="/callback?code=e2a96d3d24621b22d11f5feae2b2234630a1785cc61f6e7391579e203d74c3b1"
```

When I look at that token in the database, I see that it was created yesterday and revoked yesterday as well:

``` 
#<Doorkeeper::AccessGrant id: 57260, resource_owner_id: 22196, application_id: 174, token: "e2a96d3d24621b22d11f5feae2b2234630a1785cc61f6e7391...", expires_in: 600, redirect_uri: "https://example.com/callback", created_at: "2016-09-16 23:53:05", revoked_at: "2016-09-16 23:53:06", scopes: "profile contact events">
```

Why would doorkeeper send users an expired or revoked token?

**Comment by heavysixer** [comment]
We are seeing the same problem.

**Comment by theycallmeswift** [comment]
Is there a reason that method isn't using the `find_or_create_for` method right above it? https://github.com/doorkeeper-gem/doorkeeper/blob/ddcf2edd6045f2d54935865aed0a668b7fd03bcf/lib/doorkeeper/models/access_token_mixin.rb#L71

**Comment by heavysixer** [comment]
@theycallmeswift it actually starts there and then eventually winds up at the method i pointed  out I believe.

**Comment by theycallmeswift** [comment]
@heavysixer are you seeing this with access tokens or auth codes?  We're seeing it for auth codes which uses a different mixin.

**Comment by heavysixer** [comment]
@theycallmeswift Doorkeeper is returning expired `oauth_access_tokens` from that table. I think that means that they are access_tokens.

**Comment by theycallmeswift** [comment]
@heavysixer yeah, sounds like we're encountering different issues then.  My issue is related to Authorization Codes being expired.  May be connected somehow though.

**Comment by nbulaj** [comment]
Hi @theycallmeswift . Have you try to update Doorkeeper to a new version? 

I take a look to the `AuthorizationsController`. If I understand you correctly, the problem is expired `AccessGrant`, so your OAuth stack uses `Doorkeeper::OAuth::CodeRequest`. And it is [always creates](https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/oauth/authorization/code.rb#L13) a new Grant record (only if @token is not already set). So I could not understand how it could be...

Could you debug the code? Or maybe provide more details.

**Comment by nbulaj** [comment]
Hi @theycallmeswift. Any news here? `last_authorized_token_for` related to `AccessToken` model, `AccessGrant`'s issued in `Doorkeeper::OAuth::Authorization::Code`, where it can't be expired or revoked.

**Comment by nbulaj** [comment]
Hi @theycallmeswift . Can you please clarify what request was invoked and returned expired code? Maybe you can provide some additional info?

**Comment by theycallmeswift** [comment]
@nbulaj at this point, I don't have any additional info to provide.  This intermittent issue still persists but has proven impossible for me to recreate.  Please let me know if there are any specific queries or steps I can take that would be helpful
